servlet/src/main/java/io/grpc/servlet/ForceTrailersServerInterceptor.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2023 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.servlet;
+
+import io.grpc.ForwardingServerCall;
+import io.grpc.ForwardingServerCallListener;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+
+class ForceTrailersServerInterceptor implements ServerInterceptor {
+  @Override
+  public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call,
+      Metadata headers,
+      ServerCallHandler<ReqT, RespT> next) {
+    ForceTrailersServerCall<ReqT, RespT> interceptedCall = new ForceTrailersServerCall<>(call);
+    try {
+      return new ForwardingServerCallListener.SimpleForwardingServerCallListener<ReqT>(next.startCall(interceptedCall, headers)) {
+        @Override
+        public void onMessage(ReqT message) {
+          try {
+            super.onMessage(message);
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        @Override
+        public void onHalfClose() {
+          try {
+            super.onHalfClose();
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        @Override
+        public void onCancel() {
+          try {
+            super.onCancel();
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        @Override
+        public void onComplete() {
+          try {
+            super.onComplete();
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        @Override
+        public void onReady() {
+          try {
+            super.onReady();
+          } catch (Throwable t) {
+            closeWithThrowable(t);
+            throw t;
+          }
+        }
+
+        private void closeWithThrowable(Throwable t) {
```

Comment:
We purposefully do not propagate exceptions to the remote. See https://grpc.github.io/grpc-java/javadoc/io/grpc/util/TransmitStatusRuntimeExceptionInterceptor.html

You can use this to send the headers, though. (Although that is technically not thread-safe, we discourage exceptions, and if one happens it is probably before the service has started work on another thread.)

Reply:
What do you suggest here ? Completely remove exception handling at the level of ForceTrailsServerInterceptor and if the user wants, he can use [TransmitStatusRuntimeExceptionInterceptor](https://grpc.github.io/grpc-java/javadoc/io/grpc/util/TransmitStatusRuntimeExceptionInterceptor.html) to ensure any status runtime exception are caught and close the call ?

Reply:
If they use TransmitStatusRuntimeExceptionInterceptor, then we don't need anything here because it will do call.close() (so you just have to choose a good ordering between the interceptors). But the normal case is the exception is propagated up to gRPC, who will then close the call. I figured for that situation you'd still want to send headers here.

Reply:
I've modify the PR. Let me know what you think.